File: sdk/python/nexus_py/realtime.py
"""
nexus_py.realtime
─────────────────────
RealtimeClient — WebSocket event subscription with an event-emitter API.

Usage:
    nexus.realtime.on("npc_state_changed", my_async_handler)
    await nexus.realtime.connect(session_id)
    ...
    await nexus.realtime.disconnect()

Handlers are async callables: `async def handler(event: dict) -> None`.
Register a handler for `"*"` to receive every event regardless of type.

Connection lifecycle (mirrors the server's documented WS protocol):
    1. Open WebSocket to `{ws_url}/realtime/{session_id}`
    2. Immediately send `{"type": "auth", "token": "<access_token>"}`
    3. Server responds `{"type": "connected", ...}` on success, or closes
       with code 4001 (auth failure) / 4002 (session not found/ended)
    4. Server sends `{"type": "ping"}` every 15s — SDK auto-replies "pong"
    5. All session/NPC events arrive as JSON frames and are dispatched to
       registered handlers by their `type` (or `event` for relayed pub/sub
       frames — see `_normalise_event`)
"""

from __future__ import annotations

import asyncio
import json
import logging
from typing import TYPE_CHECKING, Any, Awaitable, Callable

import websockets
from websockets.exceptions import ConnectionClosed

if TYPE_CHECKING:
    from nexus_py.client import NexusClient

logger = logging.getLogger(__name__)

EventHandler = Callable[[dict[str, Any]], Awaitable[None]]

# Server sends a ping every 15s; reply promptly so the connection survives.
_HEARTBEAT_INTERVAL_SECONDS: float = 15.0


def _normalise_event(message: dict[str, Any]) -> dict[str, Any]:
    """
    Normalise the two wire shapes the server can send into one consistent
    dict with a `type` key, so handlers never need to branch on shape.

    WS-native frames already have `type`: {"type": "connected", ...},
    {"type": "ping"}, {"type": "npc_state_changed", "payload": {...}}.

    Pub/sub-relayed frames (session/player events forwarded from Redis)
    use `event` instead of `type`: {"event": "player.joined", "data": {...}}.
    These are remapped to {"type": "player.joined", "payload": {...}, ...}.
    """
    if "type" in message:
        return message
    if "event" in message:
        normalised = dict(message)
        normalised["type"] = normalised.pop("event")
        if "data" in normalised and "payload" not in normalised:
            normalised["payload"] = normalised.pop("data")
        return normalised
    return message


class RealtimeClient:
    """
    WebSocket event subscription client with an event-emitter pattern.

    Not typically instantiated directly — accessed via `nexus.realtime`
    on a `NexusClient`, which wires it to the client's access token and
    WebSocket base URL automatically.
    """

    def __init__(self, client: "NexusClient") -> None:
        self._c = client
        self._handlers: dict[str, list[EventHandler]] = {}
        self._ws: websockets.WebSocketClientProtocol | None = None
        self._running = False
        self._listen_task: asyncio.Task[None] | None = None
        self._heartbeat_task: asyncio.Task[None] | None = None

    def on(self, event_type: str, handler: EventHandler) -> None:
        """
        Register an async handler for a specific event type.

        Args:
            event_type: The event's `type` field to match, e.g.
                       "npc_state_changed", "player.joined", "connected".
                       Pass "*" to receive every event regardless of type.
            handler:    An async callable: `async def handler(event: dict) -> None`.

        Multiple handlers can be registered for the same event_type — all
        are called, in registration order, when a matching event arrives.
        """
        self._handlers.setdefault(event_type, []).append(handler)

    def off(self, event_type: str, handler: EventHandler | None = None) -> None:
        """
        Remove a previously registered handler.

        If `handler` is None, removes ALL handlers for `event_type`.
        """
        if event_type not in self._handlers:
            return
        if handler is None:
            self._handlers[event_type] = []
        else:
            self._handlers[event_type] = [
                h for h in self._handlers[event_type] if h is not handler
            ]

    async def connect(self, session_id: str) -> None:
        """
        Connect to the session's WebSocket channel and start listening.

        Performs the auth handshake automatically using the access token
        currently stored on the parent `NexusClient` (set by a prior
        `login()`, `guest()`, or `refresh()` call). Starts two background
        tasks: one to listen for and dispatch incoming events, and one to
        reply to server heartbeat pings.

        Raises:
            nexus_py.exceptions.AuthError: if no access token is set.
        """
        if self._c._access_token is None:
            from nexus_py.exceptions import AuthError
            raise AuthError(
                "Cannot connect to realtime channel without an access token. "
                "Call login() or guest() first."
            )

        ws_url = f"{self._c.ws_url}/realtime/{session_id}"
        self._ws = await websockets.connect(ws_url)
        self._running = True

        # Step 2 of the protocol: send the auth frame immediately.
        await self._ws.send(json.dumps({
            "type": "auth",
            "token": self._c._access_token,
        }))

        self._listen_task = asyncio.create_task(self._listen())
        self._heartbeat_task = asyncio.create_task(self._heartbeat())

        logger.debug("realtime.connected", extra={"session_id": session_id})

    async def _listen(self) -> None:
        """
        Background task: read frames from the WebSocket and dispatch to
        registered handlers. Automatically replies to server pings.

        Exits cleanly when the connection closes (server disconnect,
        network error, or explicit `disconnect()` call).
        """
        if self._ws is None:
            return

        try:
            async for raw in self._ws:
                try:
                    message = json.loads(raw)
                except json.JSONDecodeError:
                    logger.warning("realtime.invalid_json", extra={"raw": str(raw)[:200]})
                    continue

                event = _normalise_event(message)
                event_type = event.get("type", "")

                # Auto-reply to server heartbeat pings — keeps the connection
                # alive without requiring the consumer to handle "ping" themselves.
                if event_type == "ping":
                    await self._send_pong()

                handlers = (
                    self._handlers.get(event_type, [])
                    + self._handlers.get("*", [])
                )
                for handler in handlers:
                    try:
                        await handler(event)
                    except Exception as exc:
                        logger.error(
                            "realtime.handler_error",
                            extra={"event_type": event_type, "error": str(exc)},
                            exc_info=exc,
                        )
        except ConnectionClosed:
            logger.debug("realtime.connection_closed")
        finally:
            self._running = False

    async def _heartbeat(self) -> None:
        """
        Background task: send a periodic pong every 15s as a keepalive.

        The server's own ping triggers an immediate pong reply in `_listen`;
        this loop is a belt-and-suspenders safety net in case a ping frame
        is ever missed, matching the spec's reference implementation.
        """
        try:
            while self._running:
                await asyncio.sleep(_HEARTBEAT_INTERVAL_SECONDS)
                if self._running:
                    await self._send_pong()
        except asyncio.CancelledError:
            pass

    async def _send_pong(self) -> None:
        """Send a pong frame if the connection is open. Swallows send errors."""
        if self._ws is None:
            return
        try:
            await self._ws.send(json.dumps({"type": "pong"}))
        except ConnectionClosed:
            pass

    async def disconnect(self) -> None:
        """
        Close the WebSocket connection and stop all background tasks.

        Safe to call even if not currently connected (no-op in that case).
        Always called automatically by `NexusClient.close()` /
        `async with NexusClient(...) as nexus:` exit.
        """
        self._running = False

        for task in (self._listen_task, self._heartbeat_task):
            if task is not None and not task.done():
                task.cancel()
                try:
                    await task
                except asyncio.CancelledError:
                    pass

        if self._ws is not None:
            try:
                await self._ws.close()
            except Exception:
                pass
            self._ws = None

        self._listen_task = None
        self._heartbeat_task = None

    @property
    def is_connected(self) -> bool:
        """True if the WebSocket connection is currently open."""
        return self._ws is not None and self._running

File: sdk/javascript/src/realtime.ts
/**
 * nexus-js / realtime.ts
 * ─────────────────────────
 * RealtimeClient — WebSocket event subscription with an event-emitter API.
 *
 * Usage:
 *   nexus.realtime.on("npc_state_changed", (event) => {
 *     console.log(event.payload.behaviour);
 *   });
 *   await nexus.realtime.connect(session.id);
 *   ...
 *   nexus.realtime.disconnect();
 *
 * Uses the browser-native `WebSocket` global, so this file works unmodified
 * in browsers, Node 22+ (which ships WebSocket natively), and any bundler
 * target — no `ws` package dependency required.
 *
 * Connection lifecycle (mirrors the server's documented WS protocol):
 *   1. Open WebSocket to `${wsUrl}/realtime/${sessionId}`
 *   2. On open, immediately send `{"type": "auth", "token": accessToken}`
 *   3. Server replies `{"type": "connected", ...}` on success, or closes
 *      with code 4001 (auth failure) / 4002 (session not found/ended)
 *   4. Server sends `{"type": "ping"}` periodically — SDK auto-replies "pong"
 *   5. All events are dispatched to handlers registered via `.on(type, fn)`;
 *      a handler registered for `"*"` receives every event regardless of type
 */

import type { NexusClient } from "./client";
import type { NexusEvent, NexusEventHandler } from "./types";
import { NexusError } from "./types";

export class RealtimeClient {
  private ws: WebSocket | null = null;
  private handlers = new Map<string, NexusEventHandler[]>();
  /** True once the "connected" confirmation frame has been received. */
  private connected = false;

  constructor(private c: NexusClient) {}

  /**
   * Register a handler for a specific event type.
   *
   * Pass "*" to receive every event regardless of type. Multiple handlers
   * may be registered for the same event type — all are called, in
   * registration order, when a matching event arrives.
   *
   * Returns `this` so calls can be chained:
   *   nexus.realtime.on("a", h1).on("b", h2);
   */
  on(eventType: string, handler: NexusEventHandler): this {
    const existing = this.handlers.get(eventType) ?? [];
    this.handlers.set(eventType, [...existing, handler]);
    return this;
  }

  /**
   * Remove a previously registered handler for the given event type.
   * Returns `this` so calls can be chained.
   */
  off(eventType: string, handler: NexusEventHandler): this {
    const existing = this.handlers.get(eventType) ?? [];
    this.handlers.set(
      eventType,
      existing.filter((h) => h !== handler),
    );
    return this;
  }

  /**
   * Connect to the session's WebSocket channel and perform the auth handshake.
   *
   * The returned Promise resolves once the server confirms the connection
   * (receives the `{"type": "connected", ...}` frame), or rejects if the
   * connection fails, the auth handshake is rejected (close code 4001), or
   * the session is not found / already ended (close code 4002).
   *
   * Uses the access token currently stored on the parent NexusClient (set
   * by a prior `login()`, `guest()`, or `refresh()` call).
   */
  connect(sessionId: string): Promise<void> {
    if (this.c.accessToken === null) {
      return Promise.reject(
        new NexusError(
          "Cannot connect to realtime channel without an access token. Call login() or guest() first.",
        ),
      );
    }

    return new Promise<void>((resolve, reject) => {
      this.connected = false;
      const socket = new WebSocket(`${this.c.wsUrl}/realtime/${sessionId}`);
      this.ws = socket;

      socket.onopen = () => {
        socket.send(JSON.stringify({ type: "auth", token: this.c.accessToken }));
      };

      socket.onmessage = (rawEvent: MessageEvent<string>) => {
        let message: NexusEvent;
        try {
          message = JSON.parse(rawEvent.data) as NexusEvent;
        } catch {
          // Malformed frame — ignore rather than crashing the connection.
          return;
        }

        if (message.type === "connected" && !this.connected) {
          this.connected = true;
          resolve();
        }

        // Auto-reply to server heartbeat pings — keeps the connection alive
        // without requiring every consumer to handle "ping" themselves.
        if (message.type === "ping") {
          socket.send(JSON.stringify({ type: "pong" }));
        }

        this._dispatch(message);
      };

      socket.onerror = () => {
        const err = new NexusError("WebSocket connection failed");
        if (!this.connected) {
          reject(err);
        }
        // Errors after a successful connect are surfaced as a synthetic
        // "error" event rather than thrown, since there's no Promise left
        // to reject at that point — consumers can `.on("error", ...)`.
        this._dispatch({
          type: "error",
          payload: { message: err.message },
          timestamp: new Date().toISOString(),
        });
      };

      socket.onclose = (closeEvent: CloseEvent) => {
        if (!this.connected) {
          // Connection was rejected before the handshake completed.
          if (closeEvent.code === 4001) {
            reject(new NexusError("WebSocket authentication failed", "WS_AUTH_FAILED"));
            return;
          }
          if (closeEvent.code === 4002) {
            reject(new NexusError("Session not found or already ended", "WS_SESSION_INVALID"));
            return;
          }
          reject(new NexusError(`WebSocket closed before connecting (code ${closeEvent.code})`));
          return;
        }

        this.connected = false;
        this._dispatch({
          type: "disconnected",
          payload: { code: closeEvent.code, reason: closeEvent.reason },
          timestamp: new Date().toISOString(),
        });
      };
    });
  }

  /**
   * Close the WebSocket connection.
   *
   * Safe to call even if not currently connected (no-op in that case).
   * Does not throw.
   */
  disconnect(): void {
    if (this.ws !== null) {
      // Detach handlers before closing so a close triggered by disconnect()
      // doesn't fire a spurious "disconnected" event dispatch mid-teardown
      // for consumers who only care about server-initiated disconnects.
      this.ws.onclose = null;
      this.ws.onerror = null;
      this.ws.onmessage = null;
      this.ws.close();
      this.ws = null;
    }
    this.connected = false;
  }

  /** True if the WebSocket connection is open and the auth handshake completed. */
  get isConnected(): boolean {
    return this.connected;
  }

  /** Dispatch an event to all handlers registered for its type, plus all "*" handlers. */
  private _dispatch(event: NexusEvent): void {
    const typed = this.handlers.get(event.type) ?? [];
    const wildcard = this.handlers.get("*") ?? [];
    [...typed, ...wildcard].forEach((handler) => handler(event));
  }
}
